csharp/ql/src/semmle/code/csharp/commons/Assertions.qll

```diff
@@ -195,11 +197,60 @@ class SystemDiagnosticsContractAssertTrueMethod extends AssertTrueMethod {
   }
 }
 
+private predicate isDoesNotReturnIfAttributeParameter(Parameter p, boolean value) {
+  exists(Attribute a | a = p.getAnAttribute() |
+    a.getType() = any(SystemDiagnosticsCodeAnalysisDoesNotReturnIfAttributeClass c) and
+    a.getConstructorArgument(0).(BoolLiteral).getBoolValue() = value
+  )
+}
+
+/**
+ * A method with a parameter that is annotated with
+ * `System.Diagnostics.CodeAnalysis.DoesNotReturnIfAttribute(false)`.
+ */
+class SystemDiagnosticsCodeAnalysisDoesNotReturnIfAnnotatedAssertTrueMethod extends AssertTrueMethod {
+  SystemDiagnosticsCodeAnalysisDoesNotReturnIfAnnotatedAssertTrueMethod() {
+    this = any(Method m | isDoesNotReturnIfAttributeParameter(m.getAParameter(), false))
+  }
+
+  override int getAnAssertionIndex() {
+    exists(Parameter p |
+      this.getParameter(result) = p and isDoesNotReturnIfAttributeParameter(p, false)
+    )
+  }
+
+  override Class getExceptionClass() {
+    // The user defines the thrown exception.
+    any()
+  }
+}
+
+/**
+ * A method with a parameter that is annotated with
+ * `System.Diagnostics.CodeAnalysis.DoesNotReturnIfAttribute(true)`.
+ */
+class SystemDiagnosticsCodeAnalysisDoesNotReturnIfAnnotatedAssertFalseMethod extends AssertFalseMethod {
```

Comment:
Same comments as for `SystemDiagnosticsCodeAnalysisDoesNotReturnIfAnnotatedAssertTrueMethod`.